From 85848769d9e434d31b1366620b68e2c8bb6f4409 Mon Sep 17 00:00:00 2001
From: Avra Sengupta <asengupt@redhat.com>
Date: Wed, 28 Oct 2015 12:30:34 +0530
Subject: [PATCH 110/131] snapshot: Inherit snap-max-hard-limit from original volume

    Backport of http://review.gluster.org/#/c/12437/

A snapshot should inherit snap-max-hard-limit from the original
volume while being created and when being restored to, it should
restore the same.

Similarly a clone taken from a snapshot should inherit
snap-max-hard-limit from the snapshot.

Change-Id: If8e90e2ffc10e22086b803ac8e2638a16bcec968
BUG: 1275525
Signed-off-by: Avra Sengupta <asengupt@redhat.com>
Reviewed-on: http://review.gluster.org/12437
Tested-by: Gluster Build System <jenkins@build.gluster.com>
Reviewed-by: mohammed rafi  kc <rkavunga@redhat.com>
Tested-by: NetBSD Build System <jenkins@build.gluster.org>
Reviewed-by: Rajesh Joseph <rjoseph@redhat.com>
(cherry picked from commit 1f74a3efbd0337759878ffff5cd4ee6782ddfe3f)
Reviewed-on: https://code.engineering.redhat.com/gerrit/60677
Tested-by: Rajesh Joseph <rjoseph@redhat.com>
---
 tests/bugs/snapshot/bug-1275616.t             |   57 +++++++++++++++++++++++++
 tests/snapshot.rc                             |   16 +++++++
 xlators/mgmt/glusterd/src/glusterd-snapshot.c |    1 -
 xlators/mgmt/glusterd/src/glusterd-utils.c    |    1 +
 4 files changed, 74 insertions(+), 1 deletions(-)
 create mode 100755 tests/bugs/snapshot/bug-1275616.t

diff --git a/tests/bugs/snapshot/bug-1275616.t b/tests/bugs/snapshot/bug-1275616.t
new file mode 100755
index 0000000..cef07d6
--- /dev/null
+++ b/tests/bugs/snapshot/bug-1275616.t
@@ -0,0 +1,57 @@
+#!/bin/bash
+
+. $(dirname $0)/../../include.rc
+. $(dirname $0)/../../snapshot.rc
+
+function get_snap_brick_status()
+{
+    local snap=$1;
+
+    $CLI snapshot status $snap | grep "Brick Running" | sed 's/.*: //';
+}
+
+cleanup;
+TEST verify_lvm_version;
+TEST glusterd;
+TEST pidof glusterd;
+
+TEST setup_lvm 1
+
+TEST $CLI volume create $V0 $H0:$L1
+TEST $CLI volume start $V0
+TEST $CLI snapshot config activate-on-create enable
+
+TEST $CLI snapshot config $V0 snap-max-hard-limit 100
+TEST $CLI snapshot create snap1 $V0 no-timestamp
+
+TEST $CLI snapshot config $V0 snap-max-hard-limit 150
+TEST $CLI snapshot create snap2 $V0 no-timestamp
+
+TEST $CLI snapshot config $V0 snap-max-hard-limit 200
+TEST $CLI snapshot create snap3 $V0 no-timestamp
+EXPECT '197' snap_info_volume CLI "Snaps Available" $V0;
+
+TEST $CLI volume stop $V0
+
+# Restore the snapshots and verify the snap-max-hard-limit
+# and the Snaps Available
+TEST $CLI snapshot restore snap1
+EXPECT '98' snap_info_volume CLI "Snaps Available" $V0;
+EXPECT '100' snap_config_volume CLI 'snap-max-hard-limit' $V0
+
+TEST $CLI snapshot restore snap2
+EXPECT '149' snap_info_volume CLI "Snaps Available" $V0;
+EXPECT '150' snap_config_volume CLI 'snap-max-hard-limit' $V0
+
+EXPECT_WITHIN $PROCESS_UP_TIMEOUT "Yes" get_snap_brick_status snap3
+
+#Take a clone and verify it inherits snapshot's snap-max-hard-limit
+TEST $CLI snapshot clone clone1 snap3
+
+EXPECT '149' snap_info_volume CLI "Snaps Available" $V0;
+EXPECT '150' snap_config_volume CLI 'snap-max-hard-limit' $V0
+
+EXPECT '200' snap_info_volume CLI "Snaps Available" clone1
+EXPECT '200' snap_config_volume CLI 'snap-max-hard-limit' clone1
+
+cleanup;
diff --git a/tests/snapshot.rc b/tests/snapshot.rc
index 42f1b5c..52ad86a 100644
--- a/tests/snapshot.rc
+++ b/tests/snapshot.rc
@@ -431,6 +431,22 @@ function uss_count_snap_displayed() {
         ls $path/.snaps | wc -l
 }
 
+function snap_info_volume()
+{
+        eval local cli_index=\$$1
+        local var=$2
+        local vol=$3
+        $cli_index snapshot info volume $vol | grep "^$var" | sed 's/.*: //'
+}
+
+function snap_config_volume()
+{
+        eval local cli_index=\$$1
+        local var=$2
+        local vol=$3
+        $cli_index snapshot config $vol| grep "^$var" | sed 's/.*: //'
+}
+
 #return specific fields of xml output
 function get-cmd-field-xml()
 {
diff --git a/xlators/mgmt/glusterd/src/glusterd-snapshot.c b/xlators/mgmt/glusterd/src/glusterd-snapshot.c
index 82df5c4..c244d4f 100644
--- a/xlators/mgmt/glusterd/src/glusterd-snapshot.c
+++ b/xlators/mgmt/glusterd/src/glusterd-snapshot.c
@@ -9610,7 +9610,6 @@ gd_restore_snap_volume (dict_t *dict, dict_t *rsp_dict,
         strcpy (new_volinfo->volname, orig_vol->volname);
         gf_uuid_copy (new_volinfo->volume_id, orig_vol->volume_id);
         new_volinfo->snap_count = orig_vol->snap_count;
-        new_volinfo->snap_max_hard_limit = orig_vol->snap_max_hard_limit;
         gf_uuid_copy (new_volinfo->restored_from_snap,
                    snap_vol->snapshot->snap_id);
 
diff --git a/xlators/mgmt/glusterd/src/glusterd-utils.c b/xlators/mgmt/glusterd/src/glusterd-utils.c
index 009fb5c..dbcbcce 100644
--- a/xlators/mgmt/glusterd/src/glusterd-utils.c
+++ b/xlators/mgmt/glusterd/src/glusterd-utils.c
@@ -557,6 +557,7 @@ glusterd_volinfo_dup (glusterd_volinfo_t *volinfo,
         new_volinfo->brick_count = volinfo->brick_count;
         new_volinfo->tier_info = volinfo->tier_info;
         new_volinfo->quota_conf_version = volinfo->quota_conf_version;
+        new_volinfo->snap_max_hard_limit = volinfo->snap_max_hard_limit;
 
         dict_copy (volinfo->dict, new_volinfo->dict);
         dict_copy (volinfo->gsync_slaves, new_volinfo->gsync_slaves);
-- 
1.7.1

